### DISCUSSION TITLE
Is it possible for shared configs to access `context.settings` or expose extendible configs?

### DISCUSSION THREAD

**Discussion by SamuelT-Beslogic** [eslint/eslint#17346] [Q&A]
We have a shareable config file that we use across multiple projects with good default configuration. However, sometimes there are rules which require adding some custom per-project configuration on top of the base one. I would like if it was possible to use shared settings in eslint-config-* to extend certain configurations without having to create a custom plugin just for shared configs. So we don't have to re-write the complete configuration of a rule (which also won't benefit from any update in the shared config).

For example:
```jsonc
"simple-import-sort/imports": [
  "error",
  {
    "groups": [
      // Side effect imports.
      [
        "^\\u0000"
      ],
      // Packages.
      // Things that start with a letter (or digit or underscore), or `@` followed by a letter.
      [
        "^@?\\w"
      ],
      // Absolute imports and other imports such as Vue-style `@/foo`.
      // Anything not matched in another group.
      [
        "^"
      ],
      // Relative imports.
      // Anything that starts with a dot, src/, or our custom paths
      [
        "^@my-project", // <-- project specific setting
        "^(\\.|src/)"
      ]
    ]
  }
],
// ...
"no-autofix/@angular-eslint/template/i18n": [
  "error",
  {
    "boundTextAllowedPattern": "(MyProject|EN|FR)", // <-- custom for this project
    "ignoreTags": [
      // eslint-config-*
      "meta",
      "link",
      "title",
      "mat-icon"
    ],
    "ignoreAttributes": [
      // Plugin's base configuration
      "charset",
      "class",
      "color",
      "colspan",
      "fill",
      "formControlName",
      "height",
      "href",
      "id",
      "lang",
      "ngClass",
      "ngProjectAs",
      "routerLink",
      "src",
      "stroke",
      "stroke-width",
      "style",
      "svgIcon",
      "tabindex",
      "target",
      "type",
      "viewBox",
      "width",
      "xmlns",
      // eslint-config-*
      "role",
      "mat-form-field[appearance]",
      "mat-form-field[floatLabel]",
      "matTooltipPosition",
      "input[autocomplete]",
      "input[accept]",
      "name",
      "panelClass",
      "cdkColumnDef",
      "ngx-mat-select-search[name]",
      // cusom for this project
      "app-spinner[mode]"
    ]
  }
]
// ... and more
```

Which could be shortened to (and support updated):
```jsonc
"settings": {
  "my-shareable-config": {
    "ts-paths": ["@my-project"],
    "i18n": {
      "ignoreTags": [],
      "ignoreAttributes": [
        "app-spinner[mode]"
      ]
    }
  }
}
```

Is this only possible for plugins? I think I'd be fine using some sort of workaround solution if this is not to be supported. Or eventually learn how ESLint plugins work (maybe learn flat configs while I'm at it) and migrate my shared configs to a plugin.

**Comment by nzakas** [comment]
>  I would like if it was possible to use shared settings in eslint-config-* to extend certain configurations without having to create a custom plugin just for shared configs.

Yes, shareable configs support the `settings` key.

**Reply by SamuelT-Beslogic** [reply]
Maybe I just can't find it in the docs. This documentation shows how to access `context.settings` for the `create` method of a rule, but I'm not creating rules, just shared configs. https://eslint.org/docs/latest/extend/custom-rules#the-context-object

I'm trying to "receive" settings, not set them for other plugins.

**Reply by nzakas** [reply]
Ah sorry, I think I misunderstood what you're asking. Reading back over your original post, I'm still not quite sure what it is you're asking.

Shared configs are just objects -- you can modify them however you want. If you create them as JavaScript `.eslintrc.js`, then you can also `require` other files to share settings.

If this doesn't answer your question, please give me some more detail on what it is you're trying to accomplish.

**Reply by SamuelT-Beslogic** [reply]
I've seen some plugins be configurable like this:
```js
// (This is a eslint config file)

module.exports = {
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx"
      ],
      "extends": [
        "plugin:import/typescript"
      ],
      "env": {
        "es2021": true
      },
      "settings": {
        "import/parsers": {
          "@typescript-eslint/parser": [
            ".ts",
            ".tsx"
          ]
        },
        "import/resolver": {
          "typescript": {
            "typescript": {
              // Recommended by eslint-import-resolver-typescript
              // Always try to resolve types under `<root>@types` directory even it doesn't contain any source code, like `@types/unist`
              "alwaysTryTypes": true,
              "project": [
                "tsconfig?(.*).json",
                "projects/*/tsconfig?(.*).json"
              ]
            }
          }
        },
        // Required even with plugin:import/typescript
        "import/extensions": [
          ".ts",
          ".tsx"
        ]
      },
    },
  },
}
```

Plugins have access through that "settings" object through the `context` object which is a parameter when creating rules.

---

As a developer of a shared config, I'd like to expose some sort of configuration that other devs extending my shared config can configure. Plugins will do it by asking users to fill in https://eslint.org/docs/latest/use/configure/configuration-files#configuring-shared-settings , but as a non-plugin, is there any way I can access that or expose some sort of configuration?

If not, how would I go about letting users of my shared config extend rule configurations for rules that I don't own?

Even if I transformed my config into a plugin, how could I extend those rules without completely rewriting them with different default configs?

I could try and manually read a config file at the root of the user project by the `node_modules`, but that seems limiting and kinda hacky and probably couldn't work with overrides.
```js
// (this is a shared config file)

// This is more pseudo code, may not work as-is
try {
  const userConfigs = require("../../.eslintrc.js")
catch {
  const userConfigs = require("../../.eslintrc.json")
}
const userSettings = userConfigs.settings
// TODO: search through overrides?
// Warn if one of our setting if found in an override, cuz we can't respect per-file settings?

module.exports = {
  rules: {
	"simple-import-sort/imports": [
	  "error",
	  {
	    "groups": [
	      // Side effect imports.
	      [
	        "^\\u0000"
	      ],
	      // Packages.
	      // Things that start with a letter (or digit or underscore), or `@` followed by a letter.
	      [
	        "^@?\\w"
	      ],
	      // Absolute imports and other imports such as Vue-style `@/foo`.
	      // Anything not matched in another group.
	      [
	        "^"
	      ],
	      // Relative imports.
	      // Anything that starts with a dot, src/, or our custom paths
	      [
	        ...([userSettings.?["my-shareable-config"].?["ts-paths"]] || []),
	        "^(\\.|src/)"
	      ]
	    ]
	  }
	]
  }
}
```
